os/board/rtl8721csm/src/component/os/tizenrt/rtk_netmgr.c

```diff
@@ -83,15 +83,81 @@ static wifi_utils_scan_list_s *g_scan_list;
 static int g_scan_num;
 extern struct netdev *ameba_nm_dev_wlan0;
 
+struct ap_scan_list {
+	unsigned int channel;
+	char ssid[WIFI_UTILS_SSID_LEN + 1];
+	unsigned int ssid_length;
+	int rssi;
+	wifi_utils_ap_auth_type_e ap_auth_type;
+	wifi_utils_ap_crypto_type_e ap_crypto_type;
+};
+typedef struct ap_scan_list ap_scan_list_s;
+
+static ap_scan_list_s *saved_scan_list;
+static uint32_t scan_number = 0;
+uint32_t ap_channel;
+
+#include "freertos/wrapper.h"
+#define SCAN_TIMER_DURATION 180000
+static _timer scan_timer;
+
+static void _scan_timer_handler(void *FunctionContext)
+{
+	ndbg("scan Timer expired : release saved scan list\r\n");
+	ndbg("scan Timer expired : sizeof(ap_scan_list_s) =%d, scan_number=%d \r\n", sizeof(ap_scan_list_s), scan_number);
+
+	rtw_mfree((unsigned char *)saved_scan_list, sizeof(ap_scan_list_s) * scan_number);
+	scan_number = 0;
+	rtw_del_timer(&(scan_timer));
+}
+
+static int save_scan_list(wifi_utils_scan_list_s *p_scan_list)
+{
+
+	// If application calls scan before scan result free(before timeout), release scan list and cancel timer
+	if (scan_number) {
+		rtw_mfree((unsigned char *)saved_scan_list, sizeof(ap_scan_list_s) * scan_number);
+		rtw_cancel_timer(&(scan_timer));
+		ndbg("scan is called before timeout\r\n");
+	} else {
+		rtw_init_timer(&(scan_timer), &(scan_timer), _scan_timer_handler, &(scan_timer), "dynamic_chk_timer");
+		ndbg("Start scan timer\n");
+	}
+
+	rtw_set_timer(&(scan_timer), SCAN_TIMER_DURATION);
+
+	scan_number	 = 0;
+	saved_scan_list = (ap_scan_list_s *)rtw_zmalloc(sizeof(ap_scan_list_s) * g_scan_num);
+	if (saved_scan_list == NULL)
+		return RTW_NOMEM;
+
+    while(p_scan_list) {
+        strncpy(saved_scan_list[scan_number].ssid, p_scan_list->ap_info.ssid, p_scan_list->ap_info.ssid_length);
+        saved_scan_list[scan_number].ssid_length = p_scan_list->ap_info.ssid_length;
+		saved_scan_list[scan_number].channel = p_scan_list->ap_info.channel;
+		saved_scan_list[scan_number].rssi = p_scan_list->ap_info.rssi;
+		saved_scan_list[scan_number].ap_auth_type = p_scan_list->ap_info.ap_auth_type;
+		saved_scan_list[scan_number].ap_crypto_type = p_scan_list->ap_info.ap_crypto_type;
+		p_scan_list = p_scan_list->next;
+		scan_number++;
+	}
+
+	if (scan_number != g_scan_num)
+		ndbg("Total scan list temp = %d, g_scan = %d\n", scan_number, g_scan_num);

```

Comment:
Could you explain why this case is checked? When does this case happen? Do you need any measures to resolve this situation?

Reply:
@davidfather, `g_scan_num` is the total number of list on `p_scan_list`. We call `save_scan_list()` after the getting the final number of `g_scan_num` and before `TRWIFI_POST_SCANEVENT()`. 
Here with this condition, we are just double checking whether we have copied the whole of `p_scan_list` to `saved_scan_list` by comparing both it's number. The `g_scan_num = 0;` will also be done after `TRWIFI_POST_SCANEVENT()` inside `_free_scanlist();`. Hence we don't have to take any measures for this case.